### PR TITLE
ARROW-12571: [R][CI] Run nightly R with valgrind

### PR DESCRIPTION
**Pull request by jonkeane**

### CODE REVIEW

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

File: ci/scripts/r_valgrind.sh [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+# TODO: examples? vignettes? R CMD check with valgrind on?
+
+cat testthat.out
+
+# Check the valgrind output
+# We might also considering using the greps that LibthGBM uses:
+# https://github.com/microsoft/LightGBM/blob/fa6d356555f9ef888acf5f5e259dca958ca24f6d/.ci/test_r_package_valgrind.sh#L20-L85
+
+# No errors at all
```
Comment (pitrou) [inline ci/scripts/r_valgrind.sh:41]: You should use `valgrind --error-exitcode=1` instead of grepping.

File: ci/scripts/r_valgrind.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# ==273==    by 0x498BA9E: forcePromise (eval.c:555)
+# ==273==    by 0x4996C1B: FORCE_PROMISE (eval.c:5136)
+# ==273==    by 0x4996DD6: getvar (eval.c:5177)
+# ==273==    by 0x499DA15: bcEval (eval.c:6867)
+# ==273==  Uninitialised value was created by a stack allocation
+# ==273==    at 0x49CFAB5: gregexpr_Regexc (grep.c:2343)
+# ==273==
+if [[   "$(grep -c 'ERROR SUMMARY: 2 errors' testthat.out)" = 1 &&
+  "$(grep -c 'Conditional jump or move depends on uninitialised value' testthat.out)" = 1 &&
+  "$(grep -c 'Uninitialised value was created by a stack allocation' testthat.out)" = 1 ]]; then
```
Comment (pitrou) [inline ci/scripts/r_valgrind.sh:68]: We should use a Valgrind suppressions file instead of grepping:
https://stackoverflow.com/questions/17159578/generating-suppressions-for-memory-leaks

File: r/tests/testthat/test-compute-aggregate.R [hunk 3]
```diff
@@ -37,7 +37,8 @@ test_that("sum.Array", {
 
   floats <- c(floats, NA)
   na <- Array$create(floats)
-  expect_identical(as.numeric(sum(na)), sum(floats))
+  # expect_identical(as.numeric(sum(na)), sum(floats))
+  expect_identical(as.numeric(sum(na)), NA_real_)
```
Comment (pitrou) [inline r/tests/testthat/test-compute-aggregate.R:41]: Why these changes? If they have a specific purpose, there should be a comment explaining them.

File: ci/scripts/r_valgrind.sh [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -ex
+
+: ${R_BIN:=RDvalgrind}
+
+printenv
```
Comment (pitrou) [inline ci/scripts/r_valgrind.sh:23]: Isn't this dangerous if there are CI secrets in environment variables? Why is this needed?

**Comment by pitrou** [comment]
I'm trying this out and I see that it takes a lot of time without outputting anything:
```
+ RDvalgrind --vanilla -d 'valgrind --tool=memcheck --leak-check=full --track-origins=yes' -f testthat.R
```

Can the above command output progress while running the test suite?

File: ci/scripts/r_valgrind.sh (hunk 4 above)
Comment (jonkeane) [reply ci/scripts/r_valgrind.sh:23]: I was using it to diagnose if the `ARROW_RUNTIME_SIMD_LEVEL` var was being passed — secrets should be redacted even with something like `printenv`, but I've removed it regardless.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

File: r/tests/testthat/test-compute-aggregate.R (hunk 3 above)
Comment (nealrichardson) [inline r/tests/testthat/test-compute-aggregate.R:41]: Could also do something like (of course confirming the right way to determine a devel version). Could also add the devel check to the `skip_on_linux_cran` (and call it `skip_on_valgrind()`) if you wanted to make that narrower. 

```suggestion
  if (!grepl("devel", R.version.string)) {
    # Valgrind on R-devel confuses NaN and NA_real_
    expect_identical(as.numeric(sum(na)), sum(floats))
  }
```

File: docker-compose.yml [hunk 5]
```diff
@@ -1050,6 +1051,30 @@ services:
...
+      dockerfile: ci/docker/linux-r.dockerfile
+      args:
+        base: wch1/r-debug:latest
+        r_bin: RDvalgrind
+    environment:
+      <<: *ccache
+      # AVX512 not supported by Valgrind (similar to ARROW-9851) some runners support AVX512 and some do not
+      # so some build might pass without this setting
+      ARROW_USER_SIMD_LEVEL: "AVX2"
+      ARROW_RUNTIME_SIMD_LEVEL: "AVX2"
```
Comment (nealrichardson) [inline docker-compose.yml:1071]: Since you're using the R package's bundled build script, I think this is how you need to pass in extra args: 

```suggestion
      EXTRA_CMAKE_FLAGS: "-DARROW_USER_SIMD_LEVEL=AVX2 -DARROW_RUNTIME_SIMD_LEVEL=AVX2"
```

https://github.com/apache/arrow/blob/master/r/inst/build_arrow_static.sh#L75

File: r/tests/testthat/test-compute-aggregate.R (hunk 3 above)
Comment (jonkeane) [reply r/tests/testthat/test-compute-aggregate.R:41]: Yeah, I'll add the devel check to `skip_on_linux_cran()` (and probably rename it too, though I'm a bit hesitant to make it _so_ obvious that we are skipping the problematic tests — is that something we should worry about?).

As for these specific tests, looking at it now I wonder if we have benefit from `expect_identical(as.numeric(sum(na)), sum(floats))` versus `expect_identical(as.numeric(sum(na)), NA_real_)`? Is there a circumstance where we expect `sum(floats)` to change behavior and we would want to match it? I find `expect_identical(as.numeric(sum(na)), NA_real_)` to be easier to read at a glance and see exactly what it's testing/expecting.

File: docker-compose.yml (hunk 5 above)
Comment (pitrou) [reply docker-compose.yml:1071]: `ARROW_USER_SIMD_LEVEL` isn't a CMake option, which is the whole point. You use the same binary and change the instruction set at runtime.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

(with mimalloc as default)

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

File: docker-compose.yml (hunk 5 above)
Comment (nealrichardson) [reply docker-compose.yml:1073]: Thanks, suggestion amended

File: r/tests/testthat/helper-skip.R [hunk 6]
```diff
@@ -22,13 +22,19 @@ build_features <- c(
 )
 
 skip_if_not_available <- function(feature) {
+  if (feature == "re2" && identical(tolower(Sys.info()[["sysname"]]), "linux")) {
+    skip_on_cran()
+  }
+
```
Comment (nealrichardson) [inline r/tests/testthat/helper-skip.R:28]:
```suggestion
  if (feature == "re2") {
    # RE2 does not support valgrind (on purpose): https://github.com/google/re2/issues/177
    skip_on_valgrind()
  }

```

File: r/tests/testthat/helper-skip.R [hunk 7]
```diff
@@ -49,6 +55,18 @@ skip_if_not_running_large_memory_tests <- function() {
   )
 }
 
+skip_on_valgrind <- function() {
+  # This does not actually skip on valgrind, however it skips on CRAN when the 
+  # OS is linux + and the R versoin is development (which is where valgrind is run
+  # as of this code)
```
Comment (nealrichardson) [inline r/tests/testthat/helper-skip.R:61]:
```suggestion
  # This does not actually skip on valgrind because we can't exactly detect it.
  # Instead, it skips on CRAN when the OS is linux + and the R version is development 
  # (which is where valgrind is run as of this code)
```

File: docker-compose.yml [hunk 8]
```diff
@@ -1050,6 +1051,33 @@ services:
...
+        - ${REPO}:amd64-ubuntu-18.04-r-valgrind
+      args:
+        base: wch1/r-debug:latest
+        r_bin: RDvalgrind
+    environment:
+      <<: *ccache
+      # AVX512 not supported by Valgrind (similar to ARROW-9851) some runners support AVX512 and some do not
+      # so some build might pass without this setting
+      ARROW_USER_SIMD_LEVEL: "AVX2"
+      ARROW_RUNTIME_SIMD_LEVEL: "AVX2"
```
Comment (nealrichardson) [inline docker-compose.yml:1073]:
```suggestion
```

File: ci/scripts/r_valgrind.sh [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+source_dir=${1}/r
+
+${R_BIN} CMD INSTALL ${source_dir}
+pushd ${source_dir}/tests
+
+export TEST_R_WITH_ARROW=TRUE
+# to generate suppression files run:
+# ${R_BIN} --vanilla -d "valgrind --tool=memcheck --leak-check=full --track-origins=yes --gen-suppressions=all --log-file=memcheck.log" -f testthat.R 
+${R_BIN} --vanilla -d "valgrind --tool=memcheck --leak-check=full --track-origins=yes --error-exitcode=1 --suppressions=/${1}/ci/etc/valgrind-cran.supp" -f testthat.R | tee testthat.out
```
Comment (nealrichardson) [inline ci/scripts/r_valgrind.sh:31]: Have you confirmed that this job (in the script's final form) correctly fails if there's a valgrind failure? You could try this by making `skip_on_valgrind()` no-op and see if it fails as expected, then restore the skips.

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Thanks a lot for taking this on.

File: ci/scripts/r_valgrind.sh (hunk 9 above)
Comment (jonkeane) [reply ci/scripts/r_valgrind.sh:31]: Yes it does, though I will re-confirm that once we are close to merging

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
If ^^^ doesn't pass, there was a merge/squash error

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
We don't expect ^^^ to pass, it is intentionally broken

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
We don't expect ^^^ to pass, it is intentionally broken

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-valgrind

**Comment by jonkeane** [comment]
And ^^^ should pass now that we reverted our intentional failures.

**Review by nealrichardson — APPROVED** [review]
